File: source_code/java/Node.java
import java.util.ArrayDeque;

public class Node implements Comparable<Node> {
  State state;
  Node parent;
  String action;
  int cost;
  int depth;

  public Node(State state, Node parent, String action, 
      int cost, int depth) {
    this.state = state;
    this.parent = parent;
    this.action = action;
    this.cost = cost;
    this.depth = depth;
  }
  
  public void printSolution() {
    Node n = this;
    ArrayDeque<String> l = new ArrayDeque<String>();
    while (n.parent != null) {
      l.push(n.action);
      n = n.parent;
    }
    for(String a : l) {
      System.out.println(a);
    }
  }

  public int compareTo(Node n) {
    return this.cost - n.cost;
  }
}

File: source_code/python/node.py
class node:
    def __init__(self, state, parent, action, cost, depth):
        self.state = state
        self.parent = parent
        self.action = action
        self.cost = cost
        self.depth = depth

    def __str__(self):
        return '(' + ','.join([str(self.state), str(self.parent), str(self.action),
                str(self.cost), str(self.depth)]) + ')'

    def __repr__(self):
        return str(self)

    def __lt__(self, o):
        return self.cost < o.cost

    def printsolution(self):
        n = self
        l = []
        while n.parent:
            l.append(str(n.action))
            n = n.parent
        while l:
            print(l.pop())
